### PR TITLE
Uneven legend scale

### LINKED ISSUE #22: Add another scaling algorithm to evenly distribute the value space
Dynamic Range bins are [generated](https://github.com/nreese/enhanced_tilemap/blob/4.x/public/vislib/marker_types/base_marker.js#L330) from a d3.js quantize scale, evenly dividing the aggregation metric result space (value space) into 1, 3, or 5 bins.

This does not provide very useful binning when the value space is not evenly distributed. In such cases, the majority of aggregation grid cells will be visualized with identical styling. 

Another binning algorithm should be provided that dynamically sizes bins based on distribution. Bins with lots of geogrid-cells should be made smaller. Bins with little geogrid-cells should be made larger. The algorithm should produce uneven bin widths that do a better job of evenly distributing the value space.

Users should be able to select this binning algorithm with the `Options -> Aggregations Options -> Quantize scale type` select input. The input should have three options; dynamic - even width bins, dynamic - even distribution, and static. Please change the names of the input and options to something that makes more sense.

### PR DESCRIPTION
**Pull request by JacobBrandt**
Fixes #22 

With linear scaled legend
screen shot 2016-12-19 at 10 25 56 am

With uneven scaled legend
screen shot 2016-12-19 at 10 26 14 am

Both dynamic options now describe how it affects the legend.

### CODE REVIEW

**Comment by nreese** [comment]
Where did you get your test data from? How can I get it?

**Comment by JacobBrandt** [comment]
I used Twitter's streaming API.

**Comment by nreese** [comment]
This is really slick. I like how simple the logic is for splitting the bands.

I found one edge case where the top range is unbounded - but I think this is an artifact of having too many bands for the number of features.

image


I have a question about the number of bands. Should we adjust this logic with this PR as well? I am not sure the original way of picking the number of bands makes sense. I think it should be based on the number of features, not the range between min and max values. Maybe something like 1 feature gets one band, less than 9 features gets 3 bands, everything else gets 5 bands. What do you think?

File: public/vislib/marker_types/base_marker.js [hunk 1]
```diff
@@ -343,10 +349,37 @@ define(function (require) {
...
+          let max = 0;
+          let bands = this._legendColors.length;
+          features.forEach(function(obj) {
+            max = Math.max(max, obj.properties.value);
+            count++;
+            if(count >= featureLength/bands) {
+              count = 0;
+              ranges.push(obj.properties.value);
+            }
+          });
```
Comment (nreese) [inline public/vislib/marker_types/base_marker.js]: Is this forEach necessary? Couldn't the ranges just be selected by grabbing the values at the features[index] of range band splits?

File: public/options.html [hunk 2]
```diff
@@ -30,9 +30,21 @@
...
+          A legend scale which linearly grows from minimum to maximum range.
+        </div>
+      </div>
+      <div ng-if="vis.params.scaleType === 'Dynamic - Uneven'" class="form-group">
+        <div>
+          A legend scale that will create uneven ranges for the legend in an attempt
+          to split the map features uniformly across the ranges.  Useful when data is unevenly
+          distributed across the minimum - maximum range.
+        </div>
+      </div>
```
Comment (nreese) [inline public/options.html]: Can these be styled a little differently so they look more like tooltips?

File: public/vislib/marker_types/base_marker.js (hunk 1 above)
Comment (JacobBrandt) [inline public/vislib/marker_types/base_marker.js]: Yes, they could.  Good point.

File: public/options.html (hunk 2 above)
Comment (JacobBrandt) [inline public/options.html]: Sure thing.  I'll add an informational tooltip that changes based on item selected

**Comment by JacobBrandt** [comment]
I see the issue with the number of range bands in certain scenarios as well.  Your right it should go off of the number of features and not the range for this new scale.

File: public/vislib/marker_types/base_marker.js [hunk 3]
```diff
@@ -342,11 +348,44 @@ define(function (require) {
...
+        // A legend scale that will create uneven ranges for the legend in an attempt
+        // to split the map features uniformly across the ranges.  Useful when data is unevenly
+        // distributed across the minimum - maximum range.
+        let features = this.geoJson.features;
+        let featureLength = features.length;
+        features.sort(function(x, y) {
+          return d3.ascending(x.properties.value, y.properties.value);
+        });
+
+        let bottomCutoff = 1;
```
Comment (nreese) [inline public/vislib/marker_types/base_marker.js]: Lets use the same logic to select the color scale for linear and uneven dynamic scales.

Comment (JacobBrandt) [inline public/vislib/marker_types/base_marker.js]: OK

**Pull request merged**